### PR TITLE
fix: random sampling in ForgetRetainDataset

### LINKED ISSUE #139: Same retain samples across ranks
### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task
- [ ] My own task or dataset (give details below)

### Reproduction

Any official scripts.

### Expected behavior

During distributed training with multiple GPUs, all ranks produce identical retain loss values while forget loss values differ as expected. This indicates that all ranks are sampling the same retain data samples, which breaks the randomness assumption in the unlearning process.

The issue is located in `src/data/unlearn.py` in the `ForgetRetainDataset.__getitem__()` method:

```python
def __getitem__(self, idx):
    item = {}
    if self.anchor == "forget":
        item["forget"] = self.forget[idx]  # Sequential access - different across ranks ✓
        if self.retain:
            retain_idx = torch.randint(0, len(self.retain), (1,)).item()  # ❌ Problem here
            item["retain"] = self.retain[retain_idx]
```

**The problem**: `torch.randint()` uses the global PyTorch random number generator, which is synchronized across all ranks. This causes all ranks to generate the same sequence of random indices, leading to identical retain sample selection.

### PR DESCRIPTION
**Pull request by ZeguanXiao**
# What does this PR do?

Fixes [#139](https://github.com/locuslab/open-unlearning/issues/139)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Have you gone through the contributions [guide](../docs/contributing.md)?
- [x] Are your changes documented? Read documentation guidelines [here](../README.md#-further-documentation).

### CODE REVIEW

**Review by molereddy — CHANGES_REQUESTED** [review]
Thank you for the PR! Please see comment

**Review by molereddy — CHANGES_REQUESTED** [review]
It is not ideal to set the seed at the exact example level. This would mean we select the same retain example index sequences even if we are using a different dataset.

Since the point is that each rank must get a different seed, imo it is better to get the rank in the global seed function: https://github.com/locuslab/open-unlearning/blob/main/src/trainer/utils.py#L8

Let me know if you see any issues.

cc @Dornavineeth

**Comment by ZeguanXiao** [comment]
@molereddy Simply modifying `seed_everything()` doesn’t work, because the process group is only initialized after the Accelerator is created — that is, after the trainer is initialized (in `src/train.py`). Calling` seed_everything()` again after the trainer has been initialized can achieve the desired effect. Do you think this implementation is acceptable, or do you have a better suggestion?

**Comment by filyp** [comment]
If the goal is simply to have different idx for different ranks, there's also a simpler solution:
```python
        rank = torch.distributed.get_rank() if torch.distributed.is_initialized() else 0
        retain_idx = torch.randint(0, len(self.retain), (1,)).item()  # like before
        retain_idx = (retain_idx + rank) % len(self.retain)  # fans out the indices
```
Same for forget_idx. And that's the only code change that would be needed. Or for even more scrambling maybe instead of adding rank, add some hash of `rank + idx`. That should be as good as random, right?

**Comment by chengjiali** [comment]
Hi, is there any estimate on this?
